src/main/java/seedu/address/logic/core/WingmanParser.java

```diff
@@ -0,0 +1,132 @@
+        new ChangeModeCommandFactory()
+    );
+
+    /**
+     * The command groups that are available in the application.
+     */
+    private final List<CommandGroup> groups;
+
+    /**
+     * The list of top-level factories that produces commands.
+     */
+    private final List<CommandFactory<?>> factories;
+
+    /**
+     * Constructs a {@code WingmanParser} with the command groups given.
+     *
+     * @param commandGroups the command groups that are available in the
+     *                      application.
+     * @param factories     the top-level factories that are responsible
+     *                      for parsing of top-level features.
+     */
+    public WingmanParser(List<CommandGroup> commandGroups,
+        List<CommandFactory<?>> factories) {
+        this.groups = commandGroups;
+        this.factories = factories;
+    }
+
+    /**
+     * Constructs a {@code WingmanParser} with the default {@code
+     * COMMAND_GROUPS} and {@code  COMMAND_FACTORIES}
+     */
+    public WingmanParser() {
+        this(COMMAND_GROUPS, COMMAND_FACTORIES);
+    }
+
+    /**
+     * Tokenizes the user input into a {@code Deque} of {@code String}s. It
+     * splits the user input by whitespace. This may result in the token
+     * being unable to recover back to the original user input, i.e. if the
+     * user used different whitespace characters other than spaces. However,
+     * for the purpose of this application, this is acceptable.
+     *
+     * @param userInput the user input to tokenize.
+     * @return a {@code Deque} of {@code String}s that contains the tokens.
+     */
+    private static Deque<String> tokenize(String userInput) {
+        return new ArrayDeque<>(Arrays.asList(userInput.split("\\s+")));
+    }
+
+    /**
+     * Parses the user input and returns the corresponding command.
+     *
+     * @param operationMode the operation mode at which the command is
+     *                      executed.
+     * @param userInput     the user input to parse.
+     * @return the command that corresponds to the user input.
+     * @throws ParseException if a parsing error exists, or if no command is
+     *                        found in the parser.
+     */
+    public Command parse(OperationMode operationMode, String userInput) throws ParseException {
+        final Deque<String> tokens = tokenize(userInput);
+
+        final Optional<Command> topCommand = this.parseFactory(tokens);
+
+        if (topCommand.isPresent()) {
+            return topCommand.get();
+        }
+
+        final Optional<Command> groupedCommand =
+            this.parseGroup(operationMode, tokens);
+
+        if (groupedCommand.isPresent()) {
+            return groupedCommand.get();
+        }
+
+        throw new ParseException("No command group found for operation mode "
+                                     + operationMode);
+    }
+
+    private Optional<Command> parseGroup(OperationMode operationMode,
```

Comment:
one java doc missing

Reply:
I think private methods do not require java do., currently, it passes the style-check in CI. We can merge first, and latter on I will add the javadoc.